### PR TITLE
Add support for libStorage as a Kubernetes volume plugin.

### PR DESCRIPTION
**Pull request by vladimirvivien**
[Analytics]()
This change is to implement the libStorage storage provisioner and orchestrator framework as a Kubernetes volume plugin.  LibStorage is an open source project from EMC {code}.  One of its primary goals is provide persistent volume access to containerized applications.  The libStorage framework comes with a stable API and a robust set of storage primitives including: 
- Volume creation 
- Volume inspection 
- Mount and unmount
- Attach and detach

LibStorage offers a simple model for creating drivers for storage provisioning and orchestration.  Storage drivers use a small but versioned API to implement some or all storage primitives.  The drivers are automatically managed by the libStorage infrastructure.  LibStorage exists to provide uniform access to the sizable and diverse storage platforms found in enterprise deployments.  A libStorage volume plugin would allow Kubernetes to support many of these platforms while decoupling their implementations from the Kubernetes codebase.

The goals of the libStorage Kubernetes volume plugin are:
- Mitigate the efforts of implementing k8s volume services
- Grow the list of enterprise storage platforms available to k8s
- Provide a portable storage API to third-party storage implementers 

The libStorage volume plugin is designed to be cloud and provider agnostic.  The volume is able to handle both PVs and PVCs.  To enable libStorage as a dynamic persistent volume provisioner, use the `--enable-libstorage-provisioner` flag when launching the `kube-controller-manager`.  This will use libStorage to automatically provision volumes defined in a Persistent Volume Claim.

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/28599)

### CODE REVIEW

**Comment by eparis** [comment]
ok to test

**Comment by eparis** [comment]
@kubernetes/sig-storage

**Comment by pmorie** [comment]
Suggestion: this would probably be easier to review if it was split into multiple commits like so:

1.  API changes
2.  Generated code
3.  Volume plugin

**Comment by eparis** [comment]
godeps updates/changes in a single commit.

**Comment by vladimirvivien** [comment]
Working on breaking out auto-gen code from real code changes.

**Comment by vladimirvivien** [comment]
@pmorie I followed your suggestion and reset the feature branch and broke out the code into 3 separate commits to make it easy to review . 
* See 7f435b24bc3f4d722855b18a9c8339cf736eb417 - libStorage source changes
* See 56790fa95fe89e6b906030a69bf25b4d46777384 - Godep dependencies
* See 9005111dd2d81c35bbeae1f49ec3277161edb585 - Auto-generated sources

**Comment by vladimirvivien** [comment]
I signed it!

**Comment by vladimirvivien** [comment]
@kubernetes/sig-storage

**Comment by rootfs** [comment]
About `enable-libstorage-provisioner` option, please coordinate with #29006

File: Godeps/Godeps.json [hunk 1]
```diff
@@ -18,6 +18,11 @@
 			"Rev": "'75cd24fc2f2c2a2088577d12123ddee5f54e0675'"
 		},
 		{
+			"ImportPath": "github.com/BurntSushi/toml",
```
Comment (thockin) [inline Godeps/Godeps.json]: For a start we CAN NOT use this dep - it is licensed under the WTFPL which is not allowed.

https://news.ycombinator.com/item?id=5733050#up_5733477

Comment (cantbewong) [inline Godeps/Godeps.json]: Thanks for reviewing, I have asked the developer associated with the code using this dep to investigate alternatives

**Comment by vladimirvivien** [comment]
Please 'keep-open'.  This PR is still active.

**Comment by vladimirvivien** [comment]
@kubernetes/sig-storage @saad-ali 
This PR is ready to start the review process.  It is in a WIP phase, there are more code coming (testing, dynamic provisioning, etc), but wanted to get some feedback early in this release cycle.  Changes are grouped into 3 categories including new API type, dependencies, and volume plugin code.  

Looking forward to hear back.

**Comment by vladimirvivien** [comment]
@kubernetes/sig-storage   The build system seems to run into some GCE Firewall issues: 
```
ERROR: (gcloud.compute.firewall-rules.delete) Some requests did not succeed:
 - The resource 'projects/k8s-jkns-pr-gce/global/firewalls/e2e-gce-agent-pr-47-0-minion-e2e-gce-agent-pr-47-0-http-alt' was not found

ERROR: (gcloud.compute.firewall-rules.delete) Some requests did not succeed:
 - The resource 'projects/k8s-jkns-pr-gce/global/firewalls/e2e-gce-agent-pr-47-0-minion-e2e-gce-agent-pr-47-0-nodeports' was not found
```

**Comment by vladimirvivien** [comment]
@saad-ali, @kubernetes/sig-storge,
Hi any help on getting passed `FAILED   hack/make-rules/../../hack/verify-api-reference-docs.sh`  ? When I run the `hack/update-api-reference-docs.sh` script, it reports that everything is fixed, but the verify script still fails.  Any pointers here would be helpful.

**Comment by saad-ali** [comment]
> Hi any help on getting passed FAILED hack/make-rules/../../hack/verify-api-reference-docs.sh ? When I run the hack/update-api-reference-docs.sh script, it reports that everything is fixed, but the verify script still fails. Any pointers here would be helpful.

@vladimirvivien Try `make clean`, rebase with HEAD, and then run `hack/update-api-reference-docs.sh`

**Comment by wattsteve** [comment]
Thanks Vladimir, we appreciate the contribution. A couple of points on the approach taken:

- I think the Dynamic Provisioning enablement should use the Kubernetes beta Dynamic Provisioning implementation and not the Alpha. The README should be adapted to include an example with a Storage Class that uses the LibStorage Provisioner. I think this is important because we want to create a consistent user experience and process for dynamic provisioning via the Kubernetes control plane The idea is that once admin and users learns how to configure and use dynamic provisioning for one storage platform in Kubernetes, the process is as close to identical as possible when enabling and using dynamic provisioning of other storage platforms.

- I'm not sure why you have the environment variable boolean to enable lib storage provisioning. With the (newer and greatly improved) beta dynamic provisioning framework, if the admin doesn’t want cluster users to have access to a particular provisioner, then they do not create and offer a storage class for it. Using this approach, I think you can get rid of the need for the flag altogether.

- Anecdotal and not related to the volume plugin but... have you thought about running the LibStorage runtime on all nodes as a DaemonSet? One of the benefits to having the Kubernetes storage drivers (volume plugins and provisioners) in-tree is that the approach offers a good deployment and management user experience as all the drivers are included in the Kubelet and therefore are all already configured and running on every Kube node. Even with the LibStorage Volume Plugin in-tree, users are still left with the need to configure the LibStorage runtime on every Kube Node. If you were able to containerize the LibStorage runtime and deploy it as a DaemonSet, you'd be able to use the Kubernetes Control Plane to deploy the runtime and ensure that it stays up and running and greatly decrease the overall deployment complexity of getting LibStorage configured in Kubernetes.

**Comment by vladimirvivien** [comment]
@wattsteve Thank you for the thoughtful comments.  I will address them with their respective bullet points below.
* I am working on some code update that, I hope to push this week, will do several things including updating the code to use the Beta Dynamic Provisioner API instead of the Alpha.  The README, in `examples/volumes/libstorage` does have an example of using Storage Class (with the Alpha API) which will be updated. 
* The environment variable you mentioned is a relic from when this first started, before dynamic provisioning was introduced. This approach has been removed in subsequent code, and instead we rely on the standard Dynamic provisioning API.

* The libStorage infrastructure supports several deployment models including a client/server option, which we will use with the Kubernetes integration. As such, a libStorage server runtime (REX-Ray in this instance) acts as an intermediator between k8s and a storage backend for volume-related management operations. The server `does not` have have to be deployed on every node in a cluster (it may have been something I poorly described previously). Instead, the server can run on one or more (HA configuration) nodes.
 
 Since a libStorage provider's binary (REX-Ray) is not tightly coupled to the in-tree volume libStorage plugin, it can be deployed in any way that is conducive to proper k8s deployment practices.  Whether it's a `DeamonSet`, `Pod`, or a deployment external to k8s, we are exploring diverse methods that best fit libStorage/REX-Ray running inside Kubernetes while providing a management experience that remains faithful to the Kubernetes way.

Thanks again for your input!

**Comment by vladimirvivien** [comment]
@k8s-bot verify test this

**Comment by vladimirvivien** [comment]
@k8s-bot cvm gke e2e test this

**Comment by vladimirvivien** [comment]
@k8s-bot gci gke e2e test this

**Comment by vladimirvivien** [comment]
@saad-ali 
The latest push for this CR is ready to be reviewed.  The most important update in this push is considerable reduction in PR size.  The CI seems to be flaking out, is there a way to force a clean rebuild?  All my local verifications passed without any failures.

**Comment by saad-ali** [comment]
> @saad-ali  The latest push for this CR is ready to be reviewed. The most important update in this push is considerable reduction in PR size. The CI seems to be flaking out, is there a way to force a clean rebuild? All my local verifications passed without any failures.

Tests won't run until you rebase.

> This branch has conflicts that must be resolved

**Comment by vladimirvivien** [comment]
@saad-ali Please take a look at this PR for review.  Also, the number of files in this PR has gone down considerably (64 instead including all the generated source code).  But it is still marked `size/XXL`.

**Comment by vladimirvivien** [comment]
@k8s-bot cvm gke e2e test this

**Comment by dims** [comment]
@k8s-bot cvm gke e2e test this

**Comment by dims** [comment]
there you go @vladimirvivien

**Comment by vladimirvivien** [comment]
@k8s-bot cvm gke e2e test this

**Comment by vladimirvivien** [comment]
@dims thanks!

**Comment by vladimirvivien** [comment]
@k8s-bot unit test this

**Comment by eparis** [comment]
@k8s-bot test this

**Comment by mikedanese** [comment]
@k8s-bot bazel test this, issue kubernetes/test-infra#1152

**Comment by mikedanese** [comment]
@k8s-bot bazel test this, issue kubernetes/test-infra#1152

**Comment by saad-ali** [comment]
> @saad-ali Please take a look at this PR for review. Also, the number of files in this PR has gone down considerably (64 instead including all the generated source code). But it is still marked size/XXL.

Someone else from @kubernetes/sig-storage will have to take a look. I'm slammed at the moment with 1.5 release czar duties.

**Comment by vladimirvivien** [comment]
@k8s-bot unit test this

**Comment by vladimirvivien** [comment]
@k8s-bot kops aws e2e test this

**Comment by vladimirvivien** [comment]
@k8s-bot unit test this

**Pull request closed without merging**